src/TreeDataGrid.tsx

```diff
@@ -0,0 +1,428 @@
+        const [childGroups, childRowsCount] =
+          remainingGroupByKeys.length === 0
+            ? [childRows, childRows.length]
+            : groupRows(childRows, remainingGroupByKeys, startRowIndex + groupRowsCount + 1); // 1 for parent row
+        groups[key] = { childRows, childGroups, startRowIndex: startRowIndex + groupRowsCount };
+        groupRowsCount += childRowsCount + 1; // 1 for parent row
+      }
+
+      return [groups, groupRowsCount];
+    };
+
+    return groupRows(rawRows, groupBy, 0);
+  }, [groupBy, rowGrouper, rawRows]);
+
+  const [rows, isGroupRow] = useMemo((): [
+    ReadonlyArray<R | GroupRow<R>>,
+    (row: R | GroupRow<R>) => row is GroupRow<R>
+  ] => {
+    const allGroupRows = new Set<unknown>();
+    if (!groupedRows) return [rawRows, isGroupRow];
+
+    const flattenedRows: Array<R | GroupRow<R>> = [];
+    const expandGroup = (
+      rows: GroupByDictionary<R> | readonly R[],
+      parentId: string | undefined,
+      level: number
+    ): void => {
+      if (isReadonlyArray(rows)) {
+        flattenedRows.push(...rows);
+        return;
+      }
+      Object.keys(rows).forEach((groupKey, posInSet, keys) => {
+        // TODO: should users have control over the generated key?
+        const id = parentId !== undefined ? `${parentId}__${groupKey}` : groupKey;
+        const isExpanded = expandedGroupIds.has(id);
+        const { childRows, childGroups, startRowIndex } = rows[groupKey];
+
+        const groupRow: GroupRow<R> = {
+          id,
+          parentId,
+          groupKey,
+          isExpanded,
+          childRows,
+          level,
+          posInSet,
+          startRowIndex,
+          setSize: keys.length
+        };
+        flattenedRows.push(groupRow);
+        allGroupRows.add(groupRow);
+
+        if (isExpanded) {
+          expandGroup(childGroups, id, level + 1);
+        }
+      });
+    };
+
+    expandGroup(groupedRows, undefined, 0);
+    return [flattenedRows, isGroupRow];
+
+    function isGroupRow(row: R | GroupRow<R>): row is GroupRow<R> {
+      return allGroupRows.has(row);
+    }
+  }, [expandedGroupIds, groupedRows, rawRows]);
+
+  const rowHeight = useMemo(() => {
+    if (typeof rawRowHeight === 'function') {
+      return ({ row, type }: RowHeightArgs<R | GroupRow<R>>): number => {
+        if (isGroupRow(row)) {
+          return rawRowHeight({ type: 'GROUP', row });
+        }
+        return rawRowHeight({ row, type });
+      };
+    }
+
+    return rawRowHeight;
+  }, [isGroupRow, rawRowHeight]);
+
+  const getParentRowAndIndex = useCallback(
+    (row: R | GroupRow<R>) => {
+      const rowIdx = rows.indexOf(row);
+      for (let i = rowIdx - 1; i >= 0; i--) {
+        const parentRow = rows[i];
+        if (isGroupRow(parentRow) && (!isGroupRow(row) || row.parentId === parentRow.id)) {
+          return [parentRow, i] as const;
+        }
+      }
+
+      return undefined;
+    },
+    [isGroupRow, rows]
+  );
+
+  const rowKeyGetter = useCallback(
```

Comment:
Added a wrapper so we can get keys for group and regular row

Reply:
I wonder if this can break anything that expect `rowKeyGetter` to be defined when the user didn't actually pass one